docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1434,13 +1434,117 @@ The `ValueSet` message defines the following fields:
   `value_set` constructor call.
 
+According to the P4 specification, the type parameter of a Value Set, which
+defines the type of the expression that can be matched against the Value Set in
+a parser transition, and therefore determines the format of the members that can
+be inserted into the Value Set by the control plane, must be one of `bit<W>`,
+`tuple`, or `struct` [@P4SelectExpr]. The rest of this section looks at all 3 of
+these cases and gives an example `ValueSet` message when appropriate.
+
+1. If the type parameter is `bit<W>`, `match` will include exactly one
+   `MatchField` message, with the following fields (if a field is omitted here,
+   it means the default Protobuf value should be used):
+
+ * `id`: set to 1
+ * `bitwidth`: set to the value of `W`
+ * `match_type`: set to `EXACT`
+
+~ Begin P4Example
+@id(1) value_set<bit<8> >(4) pvs;
+select (hdr.f8) { /* ... */ }
+~ End P4Example
+~ Begin Prototext
+value_sets {
+  preamble {
+    id: 0x03000001
+    name: "pvs"
+  }
+  match {
+    id: 1
+    bitwidth: 8
+    match_type: EXACT
+  }
+  size: 4
+}
+~ End Prototext
+
+2. If the type parameter is a `tuple`, this version of P4Runtime does not
+   support runtime programming of the Value Set. If the P4Info message is
+   generated by a compiler, and the P4 program includes such a Value Set, the
+   compiler must reject the program.
+
+3. If the type parameter is a `struct`, this version of P4Runtime requires that
+   all the fields of the struct be of type `bit<W>` (where `W` can be different
+   for each field). If the P4Info message is generated by a compiler, and the P4
+   program includes such a Value Set, the compiler must reject the
+   program. If the Value Set is supported, the `match` field will include one
+   `MatchField` message for each field in the struct, with the following fields:
+
+ * `id`: must be unique with respect to the other `match` entries. If the P4Info
+   message was generated from a P4 compiler, we recommend that the IDs be
+   assigned incrementally, starting from 1, in the same order as the fields in
+   the P4 struct declaration.
+ * `name`: set to the name of the corresponding struct field.
+ * `annotations`: set to the list of P4 annotations associated with the struct
+   field.
+ * `bitwidth`: set to the value of `W` for the corresponding struct field.
+ * `match`: by default `match_type` is set to `EXACT`; the P4 programmer can
+   specify a different match type by using the `@match` annotation
+   [@P4SelectExpr].
+ * `doc`: documentation associated with the struct field.
+
+~ Begin P4Example
+struct match_t {
+  bit<8> f8;
+  @match(ternary) bit<16> f16;
+  @match(custom) bit<32> f32;
+}
+@id(1) value_set<match_t>(4) pvs;
+select ({ hdr.f8, hdr.f16, hdr.f32 }) { /* ... */ }
+~ End P4Example
+~ Begin Prototext
+value_sets {
+  preamble {
+    id: 0x03000001
+    name: "pvs"
+  }
+  match {
+    id: 1
+    bitwidth: 8
+    match_type: EXACT
+  }
```

Comment:
Each of the three `match` messages should have a `name` field.

Reply:
Same comment about the nesting of `match_type` and `other_match_type` inside of `match` messages that are themselves nested inside of the given `match` messages?

Reply:
Good catch for the missing names, I added them.